### PR TITLE
Add api keys export and import

### LINKED ISSUE #13346: Add API keys to user migrator exporter
User export/import should support API keys migration.

File: `user_metadata_export_service.rb`. Make sure that it's exported, imported, and the proper redis keys are created. Touching DB permissions is not necessary (they will get exported/imported with pg_dump).

### PR DESCRIPTION
**Pull request by amartinj**
Closes #13346

### CODE REVIEW

**Comment by amartinj** [comment]
retest this, please

File: app/services/carto/user_metadata_export_service.rb [hunk 1]
```diff
@@ -174,6 +200,14 @@ def export_search_tweet(search_tweet)
         updated_at: search_tweet.updated_at
       }
     end
+
+    def export_api_key(api_key)
+      api_key_json = Carto::Api::ApiKeyPresenter.new(api_key).to_poro
```
Comment (javitonino) [inline app/services/carto/user_metadata_export_service.rb]: This couples the exporter to the controller, which is something we want to avoid. That's the reason why all other exported models have an explicit export format, so they are not related at all with presenters. Also, it avoids the hack of having to add the missing attributes afterwards.

File: app/services/carto/user_metadata_export_service.rb [hunk 2]
```diff
@@ -201,9 +235,13 @@ def export_to_directory(user, path)
 
     def import_from_directory(path)
       user = user_from_file(path)
+      api_keys = []
+      api_keys += user.api_keys
+      user.api_keys = []
```
Comment (javitonino) [inline app/services/carto/user_metadata_export_service.rb]: Why is this needed? You are taking the api keys out of the user in order to save them later. Can you not simply do `user.save!` which saves all related objects?

File: app/services/carto/user_metadata_export_service.rb [hunk 3]
```diff
@@ -119,6 +127,22 @@ def build_search_tweet_from_hash(exported_search_tweet)
         updated_at: exported_search_tweet[:updated_at]
       )
     end
+
+    def build_api_key_from_hash(api_key_hash)
+      Carto::ApiKey.new(
```
Comment (javitonino) [inline app/services/carto/user_metadata_export_service.rb]: This is going to break after merging the public api key thing :P

File: app/services/carto/user_metadata_export_service.rb [hunk 4]
```diff
@@ -86,6 +91,9 @@ def build_user_from_hash(exported_user)
 
       user.layers = build_layers_from_hash(exported_user[:layers])
 
+      api_keys = exported_user[:api_keys] || []
```
Comment (javitonino) [inline app/services/carto/user_metadata_export_service.rb]: Please, bump the export format version and add a comment explaining what was added. Also, add a test to ensure that the new importer (1.0.2) can import the old version of data (1.0.1), which would basically test that this fallback to empty array works.

Something similar to: https://github.com/CartoDB/cartodb/blob/master/spec/services/carto/visualizations_export_service_2_spec.rb#L657

File: app/models/carto/api_key.rb [hunk 5]
```diff
@@ -106,6 +106,7 @@ def table_permissions_from_db
     end
 
     def create_token
+      return if @skip_token_creation
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: I prefer to put this checks at the hook level, e.g: `before_create :create_token unless: :skip_token_creation`

That way, you can still call the method directly if needed, and can easily see at a glance what hooks are being called without having to go to the hook body.

Additionally, for this particular instance, I would just skip the hook if the token is set (not nil), and avoid the `@skip_token_creation` variable.

File: app/models/carto/api_key.rb [hunk 6]
```diff
@@ -121,12 +122,28 @@ def valid_name_for_type
       end
     end
 
+    def skip_role_setup=(skip)
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: Maybe use `attr_writer` / `attr_accessor` instead of manually writing setters?

File: app/models/carto/api_key.rb [hunk 7]
```diff
@@ -160,6 +177,7 @@ def current_user
     end
 
     def create_db_config
+      return if @skip_db_config
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: Similar as above, we can skip this if the role+password is already set.

File: app/models/carto/api_key.rb [hunk 8]
```diff
@@ -172,7 +190,7 @@ def create_db_config
     end
 
     def setup_db_role
-      return if master?
+      return if master? || @skip_role_setup
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: We could skip this if the role already existed inside the database, but I'm pretty sure that may be too much.

File: app/services/carto/user_metadata_export_service.rb (hunk 3 above)
Comment (amartinj) [reply app/services/carto/user_metadata_export_service.rb]: I know. But merging that other branch would make it much harder to see this diff :P

File: app/services/carto/user_metadata_export_service.rb (hunk 4 above)
Comment (amartinj) [reply app/services/carto/user_metadata_export_service.rb]: There's already a test covering this: https://github.com/CartoDB/cartodb/blob/master/spec/services/carto/user_metadata_export_service_spec.rb#L91

Comment (javitonino) [reply app/services/carto/user_metadata_export_service.rb]: Right, but there is no test that covers only the import for api keys. In our exporter/importers the testing strategy when adding new stuff is to add the new stuff (apikeys in this case) to the [import fixture](https://github.com/CartoDB/cartodb/blob/master/spec/services/carto/user_metadata_export_service_spec.rb#L257) and then do a backwards compatibility check, that modifies the fixture (removes apikeys) before importing.

File: app/models/carto/api_key.rb [hunk 9]
```diff
@@ -109,6 +111,21 @@ def self.create_regular_key!(user: Carto::User.find(scope_attributes['user_id'])
...
+        created_at: api_key_hash[:created_at],
+        db_password: api_key_hash[:db_password],
+        db_role: api_key_hash[:db_role],
+        name: api_key_hash[:name],
+        token: api_key_hash[:token],
+        type: api_key_hash[:type],
+        updated_at: api_key_hash[:updated_at],
+        grants: api_key_hash[:grants],
+        user_id: api_key_hash[:user_id],
+        skip_role_setup: api_key_hash[:skip_role_setup]
```
Comment (javitonino) [inline app/models/carto/api_key.rb]: I would rename this method to `new_from_import` and hardcode `skip_role_setup` to `true`. This way, we keep the hack to skip this hook completely inside the model.

As a bonus, it discourages users from using this as a constructor (this is basically the same as `new`), because it forces to either us a factory, or your hooks will be skipped.

Comment (amartinj) [reply app/models/carto/api_key.rb]: Agree

**Comment by javitonino** [comment]
🔴 

STR:
- Create a user (with auth_api FF active)
- Create three untitled_tables
- Create a regular API Key `u.api_keys.create_regular_key!(name: 'wadus', grants: [{type:'apis',apis:['sql','maps']}, {type:'database', tables:[{schema:'public', name:'untitled_table', permissions:['select']}, {schema:'public', name:'untitled_table_2', permissions:['select', 'insert', 'delete', 'update']}]}])`
- Migrate the user

It fails during `pg_restore`, with:
`pg_restore: [archiver (db)] could not execute query: ERROR:  role "jtorres_ded13_role_7677cb6592b14c67d4ac99c79bf0c8b6" does not exist`

Basically, `pg_dump` is not exporting the roles (since they are per-server, not per-database), but it exports the grants (since those are per-table, which is a database object). When trying to restore the grants, it 💥 because the role does not exist.

File: services/user-mover/import_user.rb [hunk 10]
```diff
@@ -426,6 +426,15 @@ def rollback_metadata(file)
...
       # Disabling it may be hard. Maybe it's easier to just exclude it in the export.
       def import_pgdump(dump)
+        @logger.info("Creating roles for regular API Keys")
+        begin
+          Carto::User.find(@pack_config['user']['id']).api_keys.select(&:regular?).each do |k|
+            k.role_creation_queries.each { |q| superuser_pg_conn.query(q) }
+          end
+        rescue ActiveRecord::RecordNotFound
+          @logger.error("Unable to create roles for user's api keys, User id: #{@pack_config['user']['id']}")
+        end
```
Comment (javitonino) [inline services/user-mover/import_user.rb]: I'm a bit afraid to add this just because some tests of the old user mover part fail. For our current usage, this should not be happening. However, we are not going to be getting into this mess right now. 

But I would ask you to add a comment, or even better, a Rollbar message. So if this actually happens in production, we can take a look why. We are not going to look at the logger error of a migration (just the few last lines) so better to have it somewhere more visible.

Comment (javitonino) [reply services/user-mover/import_user.rb]: Also, if you could take the change to merge, it would be great!

**Comment by javitonino** [comment]
I did the same thing and it failed during metadata import with:
`2018-02-14 13:32:10 UTC: Importing: #<ActiveRecord::RecordInvalid: Validation failed: Api keys is invalid>`

**Comment by javitonino** [comment]
Never mind. The issue was that the user was broken (master api key didn't match user model). Now it's all 🍏
